### PR TITLE
`Communication`: Create channels for lectures, exercises and exams automatically

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

When lectures, exercises or exams are created, there is no channel for them created automatically.

### Description

When a lecture/exam is created/imported, or a course (not exam) exercise is created/imported, a channel is created. Therefore, we had to introduce a new relation between channels and lectures/exercises/exams in the database. It is a 1:1 relation between channel and lecture/exercise/exam. Channel now has three new columns: lecture_id, exercise_id and exam_id

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Log in to Artemis with Instructor
2. Navigate to Course Administration
3. Select Course
4. Create lecture
5. Check in messages tab, whether a channel was created 
6. Edit channel name on lecture edit page
7. Check in messages tab, whether the channel name was updated
8. Import Lecture
9. Repeat step 5
10. Repeat 4-9 for each exercise type
11. Repeat 4-9 for a real exam
12. Create test exam
13. Check that no channel was created
14. Edit test exam, check that there is no input field to edit the channel name
15. Edit old existing lecture, that already existed before deploying this branch
16. Check that there is no input field to edit the channel name
17. Repeat 15-16 for old exercises for all exercise types
18. Repeat 15-16 for old real exams.

For further testing steps, see #6548 and  #6701.

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [exam.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/entities/exam.model.ts.html) | 100% | ✅ |
| [exercise.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/entities/exercise.model.ts.html) | 89.61% | ✅ |
| [lecture.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/entities/lecture.model.ts.html) | 100% | ✅ |
| [channel.model.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/entities/metis/conversation/channel.model.ts.html) | 90.9% | ✅ |
| [exam-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/exam/manage/exams/exam-update.component.ts.html) | 95.68% | ✅ |
| [file-upload-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/exercises/file-upload/manage/file-upload-exercise-update.component.ts.html) | 84.84% | ✅ |
| [modeling-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/exercises/modeling/manage/modeling-exercise-update.component.ts.html) | 85.71% | ✅ |
| [programming-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/programming-exercise-update.component.ts.html) | 80% | ✅ |
| [programming-exercise-information.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/update/update-components/programming-exercise-information.component.ts.html) | 100% | ✅ |
| [quiz-exercise-detail.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/exercises/quiz/manage/quiz-exercise-detail.component.ts.html) | 88.27% | ✅ |
| [text-exercise-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/exercises/text/manage/text-exercise/text-exercise-update.component.ts.html) | 86.27% | ✅ |
| [lecture-update.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/lecture/lecture-update.component.ts.html) | 90.8% | ✅ |
| [title-channel-name.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Client-Tests/app/shared/form/title-channel-name/title-channel-name.component.ts.html) | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [Exercise.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/Exercise.html) | 92% | ✅ |
| [Lecture.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/Lecture.html) | 100% | ✅ |
| [Exam.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.exam/Exam.html) | 92% | ✅ |
| [Channel.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain.metis.conversation/Channel.html) | 100% | ✅ |
| [LectureRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository/LectureRepository.html) | 75% | ✅ |
| [ChannelRepository.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.repository.metis.conversation/ChannelRepository.html) | 56% | ✅ |
| [FileUploadExerciseImportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/FileUploadExerciseImportService.html) | 100% | ✅ |
| [ModelingExerciseImportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/ModelingExerciseImportService.html) | 96% | ✅ |
| [QuizExerciseImportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/QuizExerciseImportService.html) | 87% | ✅ |
| [TextExerciseImportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service/TextExerciseImportService.html) | 100% | ✅ |
| [ExamImportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamImportService.html) | 98% | ✅ |
| [ChannelService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ChannelService.html) | 93% | ✅ |
| [ConversationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.metis.conversation/ConversationService.html) | 94% | ✅ |
| [ProgrammingExerciseImportBasicService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseImportBasicService.html) | 97% | ✅ |
| [ProgrammingExerciseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseService.html) | 93% | ✅ |
| [ExamResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ExamResource.html) | 96% | ✅ |
| [FileUploadExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/FileUploadExerciseResource.html) | 100% | ✅ |
| [LectureResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/LectureResource.html) | 97% | ✅ |
| [ModelingExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ModelingExerciseResource.html) | 98% | ✅ |
| [ProgrammingExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/ProgrammingExerciseResource.html) | 90% | ✅ |
| [QuizExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/QuizExerciseResource.html) | 93% | ✅ |
| [TextExerciseResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/TextExerciseResource.html) | 99% | ✅ |
| [ChannelResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS4795-110/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.metis.conversation/ChannelResource.html) | 92% | ✅ |

### Screenshots

When creating/importing/editing a lecture/exam/exercises, there is a new input field labeled with "Channel Name". The following screenshot shows an example for programming exercises:
new input field

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/domain/Course.java [hunk 1]
```diff
@@ -196,7 +196,7 @@ public class Course extends DomainObject {
 
     @OneToMany(mappedBy = "course", fetch = FetchType.LAZY)
     @Cache(usage = CacheConcurrencyStrategy.NONSTRICT_READ_WRITE)
-    @JsonIgnoreProperties("course")
+    @JsonIgnoreProperties(value = "course")
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/Course.java:199]:
```suggestion
    @JsonIgnoreProperties("course")
```

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java [hunk 2]
```diff
@@ -183,6 +183,12 @@ public abstract class Exercise extends BaseExercise implements LearningObject {
     @Transient
     private Long numberOfRatingsTransient;
 
+    /**
+     * It is defined here for transferring the value from client.
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:187]: Shouldn't this be Part of a DTO? Instead of adding an additional attribute here that is only used in one rare case?
```suggestion
     * Used for receiving the value from client.
```

File: src/main/java/de/tum/in/www1/artemis/service/FileUploadExerciseImportService.java [hunk 3]
```diff
@@ -40,7 +45,13 @@ public FileUploadExerciseImportService(ExampleSubmissionRepository exampleSubmis
     public FileUploadExercise importFileUploadExercise(final FileUploadExercise templateExercise, FileUploadExercise importedExercise) {
         log.debug("Creating a new Exercise based on exercise {}", templateExercise);
         FileUploadExercise newExercise = copyFileUploadExerciseBasis(importedExercise);
-        return fileUploadExerciseRepository.save(newExercise);
+
+        FileUploadExercise newFileUploadExercise = fileUploadExerciseRepository.save(newExercise);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/FileUploadExerciseImportService.java:49]: Can we move this somewhere into a common method? It's currently duplicated for all exercise types.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java [hunk 4]
```diff
@@ -391,4 +406,25 @@ public List<Conversation> findAllConversationsForNotifications(User user, boolea
...
+     * @return A stream of channels for lectures/exercises that have been released
+     */
+    public Stream<Channel> filterVisibleChannelsForStudents(Stream<Channel> channels) {
+        ZonedDateTime now = ZonedDateTime.now();
+        return channels.filter(channel -> {
+            if (channel.getExercise() != null) {
+                return channel.getExercise().getReleaseDate() == null || channel.getExercise().getReleaseDate().isBefore(now);
+            }
+            else if (channel.getLecture() != null) {
+                return channel.getLecture().getStartDate() == null || channel.getLecture().getStartDate().isBefore(now);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:423]: I think lectures are still visible if the start date is in the future, so this should not be filtered here.

Also I think the exam visibility date needs to be checked for exam-channels as well.

**Review by rriyaldhi — APPROVED** [review]
Tested the following scenarios on TS4 and works as intended.
- Lecture Creation
- Lecture Import
- Exam Creation

File: src/main/java/de/tum/in/www1/artemis/domain/Exercise.java (hunk 2 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/domain/Exercise.java:187]: We have agreed with Stephan that this is ok in this case.

File: src/main/java/de/tum/in/www1/artemis/service/FileUploadExerciseImportService.java (hunk 3 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/FileUploadExerciseImportService.java:49]: I moved it into the createExerciseChannel method.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java (hunk 4 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:423]: You are right. I removed the filter for lectures and added the start date filter for exams instead.

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
When importing an exam I'm able to set an change a channel name, but no channel is created. It works for creating new exams.


Adding Exercises to Exams allows for selecting a channel name, but exam exercises should not have a channel: 
grafik

**Review by tobias-lippert — CHANGES_REQUESTED** [review]
The same as @Strohgelaender  mentioned also occurs for test exams

**Review by rriyaldhi — APPROVED** [review]
Tested (New & Old) Lecture (Creation, Editing, and Importing) and (New & Old & Test) Exam (Creation, Editing) on TS2.

**Review by MaximilianAnzinger — DISMISSED** [review]
Tested on TS2 during Testing Session. Works as described.

**Review by tobias-lippert — DISMISSED** [review]
Channels creation works for creation/import of programming exercise and text exercise

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java [hunk 5]
```diff
@@ -240,4 +248,131 @@ public void unarchiveChannel(Long channelId) {
...
+     * @param lecture     the lecture to create the channel for
+     * @param channelName the name of the channel
+     * @return the created channel
+     */
+    public Channel createLectureChannel(Lecture lecture, @NotNull String channelName) {
+        Channel channelToCreate = new Channel();
+        channelToCreate.setName(channelName);
+        channelToCreate.setIsPublic(true);
+        channelToCreate.setIsAnnouncementChannel(false);
+        channelToCreate.setIsArchived(false);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:263]: Do these values need to be specified here everytime? Aren't there some usable default values?

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java [hunk 6]
```diff
@@ -240,4 +248,131 @@ public void unarchiveChannel(Long channelId) {
...
+        if (!exercise.isCourseExercise()) {
+            return null;
+        }
+
+        Channel channelToCreate = new Channel();
+        channelToCreate.setName(channelName);
+        channelToCreate.setIsPublic(true);
+        channelToCreate.setIsAnnouncementChannel(false);
+        channelToCreate.setIsArchived(false);
+        channelToCreate.setDescription(String.format("Channel for %s exercise: %s", exercise.getExerciseType().getExerciseTypeAsReadableString(), exercise.getTitle()));
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:286]: Idea for a potential followup: Don't use a fixed string here but generate this in the client. Would allow for
- translating this
- adding an additional description while keeping the linking
- adding some direct links for quickly navigating between channel and exercise.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java [hunk 7]
```diff
@@ -240,4 +248,131 @@ public void unarchiveChannel(Long channelId) {
...
+    }
+
+    /**
+     * Update the channel of an exercise
+     *
+     * @param originalExercise the original exercise
+     * @param updatedExercise  the updated exercise
+     * @return the updated channel
+     */
+    public Channel updateExerciseChannel(Exercise originalExercise, Exercise updatedExercise) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:336]: I think we should also update the description if the exercise / exam / lecture name gets changed to maintain consistency.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java [hunk 8]
```diff
@@ -402,4 +417,25 @@ public List<Conversation> findAllConversationsForNotifications(User user, boolea
...
+     * Filter all channels where the attached lecture/exercise has been released
+     *
+     * @param channels A stream of channels
+     * @return A stream of channels for lectures/exercises that have been released
+     */
+    public Stream<Channel> filterVisibleChannelsForStudents(Stream<Channel> channels) {
+        ZonedDateTime now = ZonedDateTime.now();
+        return channels.filter(channel -> {
+            if (channel.getExercise() != null) {
+                return channel.getExercise().getReleaseDate() == null || channel.getExercise().getReleaseDate().isBefore(now);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:431]:
```suggestion
                return channel.getExercise().isReleased();
```

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java [hunk 9]
```diff
@@ -402,4 +417,25 @@ public List<Conversation> findAllConversationsForNotifications(User user, boolea
...
+            }
+            else if (channel.getExam() != null) {
+                return channel.getExam().getStartDate() == null || channel.getExam().getStartDate().isBefore(now);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ConversationService.java:434]:
```suggestion
                return channel.getExam().isVisibleToStudents();
```

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java [hunk 10]
```diff
@@ -419,13 +425,21 @@ public ProgrammingExercise updateProgrammingExercise(ProgrammingExercise program
...
         ProgrammingExercise savedProgrammingExercise = programmingExerciseRepository.save(updatedProgrammingExercise);
 
         participationRepository.removeIndividualDueDatesIfBeforeDueDate(savedProgrammingExercise, programmingExerciseBeforeUpdate.getDueDate());
         programmingExerciseTaskService.updateTasksFromProblemStatement(savedProgrammingExercise);
         // TODO: in case of an exam exercise, this is not necessary
         scheduleOperations(updatedProgrammingExercise.getId());
         groupNotificationScheduleService.checkAndCreateAppropriateNotificationsWhenUpdatingExercise(programmingExerciseBeforeUpdate, savedProgrammingExercise, notificationText);
+
+        if (channel != null) {
+            savedProgrammingExercise.setChannelName(channel.getName());
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:440]: Why is this needed? The channel got already update above and the value won't get saved.

Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:440]: Otherwise, the savedProgrammingExercise object does not include the channelName. We have to check channel for null, since it is null for exam exercises.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java (hunk 7 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:336]: We decided to remove the default description for now completely, because setting the description is an instructor's decision anyway.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java (hunk 6 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:286]: We will keep this in mind for a follow up, if needed. Good idea.

File: src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java (hunk 5 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/metis/conversation/ChannelService.java:263]: Yes, we have to set these, there is no default.

**Review by rriyaldhi — DISMISSED** [review]
Reapproved after addressing comments

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java (hunk 10 above)
Comment (Strohgelaender) [reply src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java:440]: Yes, but why does this value need to be set later on? Must it be sent to the client? Because at least on the server I don't see it used.

**Review by MaximilianAnzinger — DISMISSED** [review]
Reapproved after changes

**Review by pellumbbaboci — APPROVED** [review]

**Review by pellumbbaboci — APPROVED** [review]

**Review by rriyaldhi — DISMISSED** [review]
Reapprove after test fix

File: src/test/javascript/spec/component/lecture/lecture-detail.component.spec.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+
+    beforeEach(() => {
+        fixture = TestBed.createComponent(LectureDetailComponent);
+        component = fixture.componentInstance;
+        fixture.detectChanges();
+    });
+
+    it('should create the component', () => {
+        expect(component).toBeTruthy();
+    });
```
Comment (Strohgelaender) [inline src/test/javascript/spec/component/lecture/lecture-detail.component.spec.ts:37]: Is this a test that can actually fail? This does not seem like a useful test to me.

**Review by MaximilianAnzinger — DISMISSED** [review]
Reapproved after fixes.

File: src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.ts [hunk 12]
```diff
@@ -1,4 +1,4 @@
-import { Component, Input, OnInit, ViewChild } from '@angular/core';
+import { Component, EventEmitter, Input, OnInit, Output, ViewChild } from '@angular/core';
```
Comment (Strohgelaender) [inline src/main/webapp/app/lecture/wizard-mode/lecture-update-wizard.component.ts:1]:
```suggestion
import { Component, Input, OnInit, ViewChild } from '@angular/core';
```

File: src/main/webapp/app/shared/shared.module.ts [hunk 13]
```diff
@@ -23,6 +23,7 @@ import { AssessmentWarningComponent } from 'app/assessment/assessment-warning/as
 import { JhiConnectionWarningComponent } from 'app/shared/connection-warning/connection-warning.component';
 import { LoadingIndicatorContainerComponent } from 'app/shared/loading-indicator-container/loading-indicator-container.component';
 import { LearningGoalSelectionComponent } from 'app/shared/learning-goal-selection/learning-goal-selection.component';
+import { TitleChannelNameComponent } from './form/title-channel-name/title-channel-name.component';
```
Comment (Strohgelaender) [inline src/main/webapp/app/shared/shared.module.ts:26]:
```suggestion
```

File: src/test/javascript/spec/component/lecture/lecture-detail.component.spec.ts (hunk 11 above)
Comment (lennart-keller) [reply src/test/javascript/spec/component/lecture/lecture-detail.component.spec.ts:37]: I removed the test.

**Review by Strohgelaender — APPROVED** [review]
code

**Review by pellumbbaboci — APPROVED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Reapproved after fixes.

**Review by rriyaldhi — DISMISSED** [review]

**Review by tobias-lippert — DISMISSED** [review]
Reapprove

**Review by sven0311 — DISMISSED** [review]
tested on TS2. Worked everything as expected

**Review by sven0311 — DISMISSED** [review]
retested on ts4. Still works as expected

**Review by rriyaldhi — APPROVED** [review]
Retested on TS4 and works well as described.

**Review by TimOrtel — APPROVED** [review]
Tested on test server 4. Works as expected.

**Review by pellumbbaboci — APPROVED** [review]

**Review by rriyaldhi — DISMISSED** [review]
Retested Programming Exercise (Creation & Import) on TS1 and works as expected.

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS1 during testing session. Looks good to me.

**Review by tobias-lippert — DISMISSED** [review]
Tested on ts1. Still works. Reapprove.

**Review by sven0311 — APPROVED** [review]
reapprove

**Review by pellumbbaboci — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java [hunk 14]
```diff
@@ -114,6 +114,12 @@ OR lower(user.login) = lower(:#{#searchInput}))
     @Query("SELECT user FROM User user WHERE user.isDeleted = false AND :#{#groupName} MEMBER OF user.groups")
     List<User> findAllInGroup(@Param("groupName") String groupName);
 
+    @Query("""
+            SELECT user from User user
+            JOIN user.groups
+            """)
+    Set<User> findAllInAnyGroup();
+
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:122]: Let's add a warning here that this is only supposed to be called from test classes. Calling this on production doesn't sound like a good idea. :D

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamRegistrationService.java [hunk 15]
```diff
@@ -127,6 +137,7 @@ public List<ExamUserDTO> registerStudentsForExam(Long courseId, Long examId, Lis
                     exam.addExamUser(examUser);
                 }
 
+                addStudentToExamChannel(exam, student);
```
Comment (bassner) [inline src/main/java/de/tum/in/www1/artemis/service/exam/ExamRegistrationService.java:140]: This method call the DB each time to get the exam channel, doesn't it?
So if I add 2000 students with this method, will this crease 2000 DB calls? 🤔 

Since you already have `ChannelService#deregisterUsersFromExamChannel`, I suggest to create a similar method to add register students to the exam channel as well.

File: src/main/webapp/app/overview/discussion-section/discussion-section.component.ts [hunk 16]
```diff
@@ -154,12 +162,73 @@ export class DiscussionSectionComponent extends CourseDiscussionDirective implem
...
+            this.channelService
+                .getChannelOfLecture(courseId, this.lecture.id)
+                .pipe(map((res: HttpResponse<Channel>) => res.body))
+                .subscribe(getChannel());
+        } else if (this.exercise?.id) {
+            this.channelService
+                .getChannelOfExercise(courseId, this.exercise.id)
+                .pipe(map((res: HttpResponse<Channel>) => res.body))
+                .subscribe(getChannel());
+        }
```
Comment (bassner) [inline src/main/webapp/app/overview/discussion-section/discussion-section.component.ts:217]: I'm wondering if we can avoid these additional REST calls by including them in a previous request aka the ones where the lecture or the exercise come from? 🤔

**Review by bassner — CHANGES_REQUESTED** [review]
Hi! I've had a look at your code and stumbled over 3 small things. Could you please address them? :)

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java (hunk 14 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java:122]: I suggest to move this method into a test repository then

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamRegistrationService.java (hunk 15 above)
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamRegistrationService.java:140]: Agree, please make sure that adding to the channel is done asynchronously

File: src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java (hunk 14 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/repository/UserRepository.java]: We moved the method to a test repository

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamRegistrationService.java (hunk 15 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamRegistrationService.java]: We created a new method `ChannelService#registerUsersToExamChannel` and add students to the exam channel asynchronously.

**Review by MaximilianAnzinger — DISMISSED** [review]
Tested on TS2 during testing session. Works as intended.

**Review by tobias-lippert — DISMISSED** [review]
Tested in testing session on ts2. Everything works as intended.

**Review by pellumbbaboci — DISMISSED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Tested on TS2 during testing session. Works as described.

**Review by pellumbbaboci — APPROVED** [review]

**Review by bassner — APPROVED** [review]
